apps/src/lib/script-editor/FlexGroup.jsx

```diff
@@ -0,0 +1,219 @@
+    boxShadow: 'none',
+    margin: '0 10px 10px 10px'
+  }
+};
+
+class FlexGroup extends Component {
+  static propTypes = {
+    addGroup: PropTypes.func.isRequired,
+    addStage: PropTypes.func.isRequired,
+    stages: PropTypes.array.isRequired,
+    levelKeyList: PropTypes.object.isRequired
+  };
+
+  handleAddGroup = () => {
+    this.props.addGroup(
+      prompt('Enter new stage name'),
+      prompt('Enter new group name')
+    );
+  };
+
+  handleAddStage = position => {
+    this.props.addStage(position, prompt('Enter new stage name'));
+  };
+
+  /**
+   * Generate the ScriptDSL format.
+   * @param stages
+   * @return {string}
+   */
+  serializeStages = stages => {
+    let s = [];
+    stages.forEach(stage => {
+      let t = `stage '${stage.name}'`;
+      if (stage.lockable) {
+        t += ', lockable: true';
+      }
+      if (stage.flex_category) {
+        t += `, flex_category: '${stage.flex_category}'`;
+      }
+      s.push(t);
+      stage.levels.forEach(level => {
+        if (level.ids.length > 1) {
+          s.push('variants');
+          level.ids.forEach(id => {
+            s = s.concat(this.serializeLevel(id, level, level.activeId === id));
+          });
+          s.push('endvariants');
+        } else {
+          s = s.concat(this.serializeLevel(level.ids[0], level));
+        }
+      });
+      s.push('');
+    });
+    return s.join('\n');
+  };
+
+  serializeLevel(id, level, active) {
+    if (id === NEW_LEVEL_ID) {
+      return;
+    }
+
+    const s = [];
+    const key = this.props.levelKeyList[id];
+    if (/^blockly:/.test(key)) {
+      if (level.skin) {
+        s.push(`skin '${level.skin}'`);
+      }
+      if (level.videoKey) {
+        s.push(`video_key_for_next_level '${level.videoKey}'`);
+      }
+      if (level.concepts) {
+        s.push(`concepts ${level.concepts}`);
+      }
+      if (level.conceptDifficulty) {
+        s.push(`level_concept_difficulty '${level.conceptDifficulty}'`);
+      }
+    }
+    let l = `${this.normalizeLevelKind(level.kind)} '${key.replace(
+      /'/,
+      "\\'"
+    )}'`;
+    if (active === false) {
+      l += ', active: false';
+    }
+    if (level.progression) {
+      l += `, progression: '${level.progression}'`;
+    }
+    s.push(l);
+    return s;
+  }
+
+  /**
+   * Levels with kind "puzzle" and "unplugged" are special cases of "level", for
+   * the purpose of the ScriptDSL.
+   * @param kind
+   * @return {string}
+   */
+  normalizeLevelKind(kind) {
+    return !kind || kind === LevelKind.puzzle || kind === LevelKind.unplugged
+      ? LevelKind.level
+      : kind;
+  }
+
+  render() {
+    const groups = _.groupBy(
+      this.props.stages,
+      stage => stage.flex_category || 'Default'
+    );
+    let count = 0;
+    let afterStage = 1;
+
+    return (
+      <div>
+        {_.map(groups, (stages, group) => (
+          <div key={group}>
+            <div style={styles.groupHeader}>
+              Group {++count}: {group}
```

Comment:
Yuck :(

Reply:
This one I've fixed. Everything else is captured in the linked doc. Thanks for the detailed review!